### DISCUSSION TITLE
Is there any way to get Git shortlog output, but sorted by most recent commit?

### DISCUSSION THREAD

**Discussion by eamodio** [git-for-windows/git#3067] [Q&A]
As far as I can see git can only sort the output of `shortlog` alphabetically or by commit count.

**Comment by dscho** [comment]
Reading https://git-scm.com/docs/git-shortlog, I would agree.

And reading the source code at https://github.com/git/git/blob/v2.30.1/builtin/shortlog.c#L449-L451, I am certain.

Are there other ways you would want to sort the shortlog?

**Reply by eamodio** [reply]
Sorting by most recent commit, is the only request I've heard (other than what is already supported).

**Reply by dscho** [reply]
Could you show me a mock up of how you want it to look like?

**Reply by eamodio** [reply]
Thinking about this more, I think the actual sorting is less important (though nice), than including the date of the most recent commit for that author.

Currently I'm using `git shortlog -sne --all --no-merges HEAD`
![image](https://user-images.githubusercontent.com/641685/109747820-d3e8c600-7ba5-11eb-926f-d80cb18c6a68.png)

But it would be great if the most recent commit date could be included in that output, e.g. a `--date` flag. And if the sorting was also added a `--recent` flag?

I am using that info to power the _Contributors_ view in GitLens:
![image](https://user-images.githubusercontent.com/641685/109747627-7eacb480-7ba5-11eb-96d5-29f69795cbb6.png)

So I use `shortlog` to provide the set of contributor and counts, and then when you expand each contributor it shows the `git log` for them. But since largest number of commits isn't always a good indicator/useful for all project, it has been requested to be able to sort the contributors by recent commits.

Related, is there any way to also get `shortlog` to work against all branches rather than just a specified revision/range?

**Reply by dscho** [reply]
Aren't we safely outside the Windows-specific realm of Git _for Windows_ now? 😉 

I think it is safe to say that now is the time to send this as a concrete proposal to [the Git mailing list](mailto:git@vger.kernel.org) (please make sure to send plain text only, even an alternate HTML part will make sure that the mail gets dropped).

**Reply by jasonwilliams** [reply]
@dscho do you have (or a link to) any proposals sent to that mailing list that may be used as an example?

**Reply by dscho** [reply]
@jasonwilliams I think https://lore.kernel.org/git/20200925070120.GA3669667@coredump.intra.peff.net/ is a very good example to follow. It presents the overall goal clearly in the cover letter, and then "tells a story" by way of 8 patches.

You will also see that there has been quite some review on the first iteration of the patch series. As a consequence, the author reworked the patches (using [the interactive rebase](https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history), see also [this description, targeting Git Kraken users](https://blog.axosoft.com/learning-git-interactive-rebase/)) and sent off the second iteration of the patch series, which eventually made it into an official Git version.

If the mechanics of contributing patch series to a mailing list seem daunting, you need not worry, there is [GitGitGadget](https://gitgitgadget.github.io/) to help you: just open a PR at https://github.com/git/git/ or https://github.com/gitgitgadget/git/ and it will guide you through the process.

If you don't know where to start in the code, I would recommend reading the commits corresponding to above-mentioned patch series (you can see them here: https://github.com/gitgitgadget/git/compare/jk/shortlog-group-by-trailer~8...jk/shortlog-group-by-trailer). That will give you an idea what parts of Git's code base need to be touched, and it will also guide you to the code that you will want to understand (and then modify).
